
project external/srec/
diff --git a/external/srec/config/en.us/baseline.par b/external/srec/config/en.us/baseline.par
index 362efc6..6e4ee25 100644
--- a/external/srec/config/en.us/baseline.par
+++ b/external/srec/config/en.us/baseline.par
@@ -20,7 +20,7 @@ cmdline.bgsniff      = 25;
 
 # beginning of speech detection stuff
 cmdline.bgsniff_min     = 4;
-cmdline.silence_duration_in_frames = 100;
+cmdline.silence_duration_in_frames = 1;
 cmdline.end_of_utterance_hold_off_in_frames = 30;
 # 1 means yes do start-pointing
 cmdline.gatedmode    = 1;
diff --git a/external/srec/config/en.us/baseline11k.par b/external/srec/config/en.us/baseline11k.par
index c580885..f24436e 100644
--- a/external/srec/config/en.us/baseline11k.par
+++ b/external/srec/config/en.us/baseline11k.par
@@ -25,7 +25,7 @@ cmdline.bgsniff_min     = 4;
 cmdline.gatedmode    = 1;
 
 # level-based end of speech detection stuff
-cmdline.silence_duration_in_frames = 100;
+cmdline.silence_duration_in_frames = 1;
 cmdline.end_of_utterance_hold_off_in_frames = 30;
 
 # logging
diff --git a/external/srec/config/en.us/baseline8k.par b/external/srec/config/en.us/baseline8k.par
index 69d6ea8..9537be3 100644
--- a/external/srec/config/en.us/baseline8k.par
+++ b/external/srec/config/en.us/baseline8k.par
@@ -25,7 +25,7 @@ cmdline.bgsniff_min     = 4;
 cmdline.gatedmode    = 1;
 
 # level-based end of speech detection stuff
-cmdline.silence_duration_in_frames = 100;
+cmdline.silence_duration_in_frames = 1;
 cmdline.end_of_utterance_hold_off_in_frames = 30;
 
 # logging
