
project hardware/samsung_slsi-cm/exynos5420/
diff --git a/hardware/samsung_slsi-cm/exynos5420/libvirtualdisplaymodule/ExynosVirtualDisplayModule.h b/hardware/samsung_slsi-cm/exynos5420/libvirtualdisplaymodule/ExynosVirtualDisplayModule.h
index 44c31ff..b9315fe 100644
--- a/hardware/samsung_slsi-cm/exynos5420/libvirtualdisplaymodule/ExynosVirtualDisplayModule.h
+++ b/hardware/samsung_slsi-cm/exynos5420/libvirtualdisplaymodule/ExynosVirtualDisplayModule.h
@@ -11,4 +11,13 @@ class ExynosVirtualDisplayModule : public ExynosVirtualDisplay {
         virtual int32_t getDisplayAttributes(const uint32_t attribute);
 };
 
+enum {
+    HWC_DISPLAY_COMPOSITION_TYPE = 0,
+    HWC_DISPLAY_GLES_FORMAT,
+    HWC_DISPLAY_SINK_BQ_FORMAT,
+    HWC_DISPLAY_SINK_BQ_USAGE,
+    HWC_DISPLAY_SINK_BQ_WIDTH,
+    HWC_DISPLAY_SINK_BQ_HEIGHT
+};
+
 #endif
