diff --git a/vendor/cm/config/common.mk b/vendor/cm/config/common.mk
index 00a536c..aedee10 100644
--- a/vendor/cm/config/common.mk
+++ b/vendor/cm/config/common.mk
@@ -251,13 +251,6 @@ PRODUCT_PROPERTY_OVERRIDES += \
     media.sf.omx-plugin=libffmpeg_omx.so \
     media.sf.extractor-plugin=libffmpeg_extractor.so
 
-# Telephony
-PRODUCT_PACKAGES += \
-    telephony-ext
-
-PRODUCT_BOOT_JARS += \
-    telephony-ext
-
 # These packages are excluded from user builds
 ifneq ($(TARGET_BUILD_VARIANT),user)
 PRODUCT_PACKAGES += \
diff --git a/config/telephony.mk b/config/telephony.mk
index 2503125..29d3e97 100644
--- a/vendor/cm/config/telephony.mk
+++ b/vendor/cm/config/telephony.mk
@@ -6,7 +6,11 @@ PRODUCT_COPY_FILES += \
 PRODUCT_PACKAGES += \
     messaging \
     Stk \
-    CellBroadcastReceiver
+    CellBroadcastReceiver \
+    telephony-ext
+
+PRODUCT_BOOT_JARS += \
+    telephony-ext
 
 # Default ringtone
 PRODUCT_PROPERTY_OVERRIDES += \
